mailbox: dynamic auth data size

ellemouton opened the pull request:
Fixes #16 

In this PR, we allow the `authData` sent by the noise responder in Act 2 to be of variable length. In order to accomplish this, the Act 2 message had to be changed slightly so that the length of the auth data could be included before the `EncryptAndHash` step so that we know how many bytes to read for that step. 

In order to read Act 2 payload in a more stream-like way, the first commit adds a recv buffer to the `connKit` so that we can call `Read` multiple times without actually calling GBN to recv multiple times

lightninglabs-deploy commented:
@guggero: review reminder

guggero commented on mailbox/noise.go:
```diff
@@ -662,33 +655,23 @@ func (b *Machine) GenActTwo() ([ActTwoSize]byte, error) {
+	// We'll write out 2 bytes that denotes the true length of the
+	// authentication payload, followed by the payload itself.
+	var payloadLen [2]byte
+	binary.BigEndian.PutUint16(payloadLen[:], uint16(len(authPayload)))
```
Any reason to go with `uint16` and not `uint32` directly? This would limit us to 65k of payload which might be small if we're adding a bunch of permissions and caveats to a macaroon.

guggero replied:
Also, we should probably use a constant for the `2` here, just so we can make this change more easily without forgetting it in some place.

guggero commented on mailbox/noise.go:
```diff
@@ -697,25 +680,47 @@ func (b *Machine) GenActTwo() ([ActTwoSize]byte, error) {
-	copy(s[:], actTwo[34:83])
-	copy(p[:], actTwo[83:])
+	// Read the payload length key.
+	var payloadLenBytes [2]byte
```
nit: move declaration to top as well?

guggero commented on mailbox/noise.go:
```diff
@@ -697,25 +680,47 @@ func (b *Machine) GenActTwo() ([ActTwoSize]byte, error) {
+	}
+
+	authPayloadLen := binary.BigEndian.Uint16(payloadLenBytes[:])
+	p := make([]byte, authPayloadLen)
```
If we're going with `uint32` we should probably still check the length here to avoid allocating up to 4GB of memory if the remote party is malicious. Perhaps we should limit the size to 1MB for now and error out if a larger length is sent here?

Roasbeef replied:
I think a `uint16` is more than enough: why would macaroons ever get that large?

ellemouton replied:
thoughts @guggero?

guggero reviewed (commented):
Great fix, thank you for tackling this so quickly!
Just a few comments around the max payload size, otherwise LGTM :tada:

guggero commented on mailbox/noise.go:
```diff
@@ -501,18 +509,14 @@ const (
+	// ActTwoOverheadSize is the minimum size of the packet sent from
+	// responder to initiator in ActTwo. The packet consists of a
+	// handshake version, an ephemeral key in compressed format, the static
+	// key of the responder (encrypted with a 16 byte MAC) and 2 byte field
```
nit: 4 bytes, here and below. Total should come to 87 now as well?

ellemouton replied:
ah yes! 🙈  thanks!

ellemouton replied:
looks like it is still missing something else aswell (itests are hanging) ... looking into it

ellemouton replied:
ok cool, found the other bug. was also constant related 🙃 updated! 🚀

Roasbeef commented on mailbox/interface.go:
```diff
@@ -108,19 +108,26 @@ type connKit struct {
-	data := NewMsgData(ProtocolVersion, nil)
-	if err := k.impl.ReceiveControlMsg(data); err != nil {
-		return 0, err
+	if k.recvBuffer.Len() == 0 {
```
What's the implication of _not_ having this? The we violate the normal `Read` semantics and return potentially _more_ bytes that the caller expected (seen via the value of the `n` param).

ellemouton replied:
yeah so this ensures that we dont violate this rule.

The reason this was not required before is that all the acts were specific expected lengths and so we knew how big the byte slice we passed in here needed to be. But now, if we have a dynamic field in the Act 2 message, we will not know how big the slice needs to be and need to first read the length field. But since the entire payload is returned from gbn in one go, we need this cache buffer here.

Roasbeef commented on mailbox/noise.go:
```diff
@@ -501,18 +509,14 @@ const (
 
-	// ActTwoPayloadSize is the size of the fixed sized payload that can be
-	// sent from the responder to the initiator in act two.
-	ActTwoPayloadSize = 500
```
Picked this initially as it seemed like one of those "good enough" values. There's also the added simplicity in keeping this fixed size as well. 

Do we see a near term future that results in the macaroon being more than 500 bytes?

Also if we want to decouple things further, then what we can do is actually just send the final payload over in a message from the server to client, but consider this to be a final act (4) of the handshake.

Roasbeef replied:
The other advantage of keeping this fixed size (even if we increase to say 1 KB or something) is traffic analysis: all handshakes look the same, and no information can be inferred w.r.t the size of the macaroon payload in the second act.

ellemouton replied:
> Do we see a near term future that results in the macaroon being more than 500 bytes?

The current admin supermacaroon is already more than 500 bytes. Was busy with the admin-macaroon session task and then notices that the macaroon was getting cut in half here which is why i had to make this change.

> Also if we want to decouple things further, then what we can do is actually just send the final payload over in a message from the server to client, but consider this to be a final act (4) of the handshake.

yeah could do this and then in act 2 just send over the size of the payload to be expected in act 4. this would then also allow us to get rid of the previous commit since we will always know how many bytes the next message will be. However, this still has the issue of traffic analysis....

ellemouton replied:
The current admin supermacaroon we bake is 1026 bytes

Roasbeef replied:
> However, this still has the issue of traffic analysis....

Yeah the main way to get around this would be to pick something "large enough" and always send that, trading off the extra bandwidth for the privacy gain. So it could be like 65 KB each time or something.

Roasbeef reviewed (changes requested):
Great initial attempt at this! 

Left some comments line re fixed vs dynamic size, and also another option re extension of the handshake itself. 

Next follow up here should be on the `litd` side, re doing away with the existing password usage and also adding a default mode that is read-only to start with. 

Other thing we need to work out is the upgrade path here, and if we want to retain the older logic for a period of time (support both v0 and v1) before removing the older handshake all together. 

Re the version bump, we could also hold off on the bump until the second portion of the handshake is complete (uses pure long term public keys)

ellemouton reviewed (commented):
Thanks for the review @Roasbeef 🚀 

I left some questions on your comments

guggero commented:
> why would macaroons ever get that large?

Macaroons can become quite big if we start using them more precisely. For example we could start pinning them to certain URIs. Or add a whole list of caveats for security restrictions implemented in the macaroon firewall. So I can easily see a macaroon growing larger than 65k (uint16) bytes.

> yeah could do this and then in act 2 just send over the size of the payload to be expected in act 4. 

I do like the idea of making the payload in act 2 static and only sending over the (uint32) size of the payload to be expected in act 4.

> Other thing we need to work out is the upgrade path here, and if we want to retain the older logic for a period of time (support both v0 and v1) before removing the older handshake all together.

I wasn't sure how/when we'd deploy this. Whether it would be a "big bang, you now need to update" or whether we always wanted to remain backward compatibility. So I told @ellemouton to not worry about compatibility for now.
But if we want to bundle this in with some other changes (e.g. the long-term mailbox ID derivation stuff) we might need to start worrying about compatibility.

@ellemouton at this point I don't think it would be a lot of code to conditionally do an act 4 based on the version number alone, right?

ellemouton commented:
> @ellemouton at this point I don't think it would be a lot of code to conditionally do an act 4 based on the version number alone, right?

cool, yeah this sounds like a good idea 👍

ellemouton commented:
Cool, updated 🚀 

With this latest update, a sneaky optional act 4 is added which is just a normal message send & receive but happens before the handshake is complete. So the act 2 payload is just used to send along the size of the authdata to be sent in act 4. 

This change allows for backwards compatibility in the case that the client (TW) is updated before the server (Litd). What we do is let the client send its lowest compatible version, in this case 0, and then the server will respond with the latest version it is compatible with. So: if the client updates to v1 and the server is on v0, the client will send v0 in act1, the server will respond with v0 in Act 2 and the client sees this and takes it to mean it should grab the auth data from act 2.
Else, if the server is on v1, it will respond with v1 in act 2 and then the client will take the payload in act 2 to mean the length of the auth data that will be sent along in act 4.

guggero commented on mailbox/noise.go:
```diff
@@ -489,10 +497,20 @@ func ekeUnmask(me *btcec.PublicKey, password []byte) *btcec.PublicKey {
+	// MinHandshakeVersion and HandshakeVersion will cause the handshake
 	// to abort immediately.
-	HandshakeVersion = byte(0)
+	HandshakeVersion = byte(1)
```
Should we define `HandshakeVersion1` above as well, then just point this to that version? And perhaps rename this to `CurrentHandshakeVersion` to make it less confusing naming wise?

guggero commented on mailbox/noise.go:
```diff
@@ -584,9 +606,13 @@ func (b *Machine) RecvActOne(actOne [ActOneSize]byte) error {
 	// TODO(roasbeef); use version to gate the initial and subsequent
 	// pairings
-	if actOne[0] != HandshakeVersion {
+	b.remoteVersion = actOne[0]
```
nit: can remove the TODO above now I think.

guggero commented on mailbox/noise.go:
```diff
@@ -761,12 +788,25 @@ func (b *Machine) RecvActTwo(actTwo [ActTwoSize]byte) error {
 
-	payloadLen := binary.BigEndian.Uint16(payload[:2])
-	b.authData = make([]byte, payloadLen)
+	switch b.remoteVersion {
```
Would be nice to add some unit tests for these version negotiations. I think adding a new functional option to create a client or server with a specific option should be pretty easy.

ellemouton replied:
do you mean that the functional option should specify the version?

and if so, but then we need to keep all the previous code too right? like for example, how will we test that the upgraded client works with the older server?

Or do you mean just create a unit test for each of the 'GenActX' functions and ensure that they return the correct data based on what is passed in?

Roasbeef replied:
Agreed re tests here, only 4 combinations here so far that we can enumerate pretty easily. 

See comment above about awkwardness of sending the length of a future message vs just using the existing length prefix system to do so. If we just accept the packet size leak here, we can simplify by using a larger length int (uint32 instead of uint16) for the payload. And we do that _only_ for handshake v1, v0 uses the old 2 byte length prefix.

guggero reviewed (commented):
Very nice that we could solve the version negotiation in this way!
Latest version looks good code wise, only think missing are some unit tests IMO.

lightninglabs-deploy commented:
@roasbeef: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@roasbeef: review reminder
@ellemouton, remember to re-request review from reviewers when ready

Roasbeef commented on mailbox/grpc_noise_conn.go:
```diff
@@ -215,6 +215,17 @@ func (c *NoiseGrpcConn) ClientHandshake(_ context.Context, _ string,
 		return nil, nil, err
 	}
 
+	// If the server has indicated that there will be authData sent in
```
Can't this just be gated based on the protocol version? So v0 is what we have rn, and then in v1 we expect a payload. The payload can just be a regular message though, and we wouldn't need to do anything special here other than do a normal call to `ReadMessage` here, as for v1 we _always_ send that extra message. 

FWIW, from a network perspective, it's more or less the same as having dynamic data be sent along with the payload in the earlier act. Apologies that I fell behind on this PR a bit, but my initial comment was mainly meant to force us to ackledge the implications of dynamic payloads.

ellemouton replied:
cool. yeah the only reason i did it like this (ie, not using `ReadMessage`) is because of the current 2 byte length prefix limit for normal messages. But if it is fine for me to change that, ie use 2 byte length prefix for v0 comms and use 4 byte length prefix for all v1 comms, then yes that makes things way simpler

Roasbeef commented on mailbox/grpc_noise_conn.go:
```diff
@@ -215,6 +215,17 @@ func (c *NoiseGrpcConn) ClientHandshake(_ context.Context, _ string,
+	// If the server has indicated that there will be authData sent in
+	// an Act Four, then read that now.
+	if c.noise.actFourSize > 0 {
+		actFour := make([]byte, c.noise.actFourSize)
```
So re the comment above, we'd just read it as a normal encrypted message w/ the length prefix.

Roasbeef commented on mailbox/noise.go:
```diff
@@ -659,30 +685,26 @@ func (b *Machine) GenActTwo() ([ActTwoSize]byte, error) {
-	// However, we also want to send the client data they may need for
-	// authentication (if present) encrypted with strong forward secrecy.
+
+	// We include the size of tha auth data that we will send along to the
```
Weird protocol flow IMO: why do we need to send the size of a message that we'll send at a future time? We can just use the normal length prefix scheme to do that, packaging this payload within the auth payload we send anyway. 

We should also only do this if we're on the _newer_ handshake version.

ellemouton commented:
Thanks for the review @Roasbeef  & @guggero 🚀 

I have updated it quite a bit now. It no longer has the weird flow where the length of act 4 is sent in act 2. Now instead, act 4 is just a normal message _but_ to do this, I had to change the normal-message payload length prefix to be 4 bytes. This only takes effect for version v1 and up though.

I also added unit tests to ensure that the negotiations between different versions happens smoothly

Roasbeef commented:
What of the suggestion to just package it with act 3? Similar to the
original version of this PR. That should result in the most minimal diff:
use the 4 byte length there, but reject the handshake if it goes over a
soft cap we set for a sane macaroon size.

It's also possible to send a message across multiple logic transport
messages as well, IIRC the current brontide code in lnd still does this.

Usage of 2 bytes for the message length is purposeful, as it means we can
always just have a 65KB buffer to read messages into, and lets us side step
some issues w/ ppl sending massive messages to DoS the client.

On Mon, Jan 10, 2022 at 5:47 AM Elle ***@***.***> wrote:

> Thanks for the review @Roasbeef <https://github.com/Roasbeef> & @guggero
> <https://github.com/guggero> 🚀
>
> I have updated it quite a bit now. It no longer has the weird flow where
> the length of act 4 is sent in act 2. Now instead, act 4 is just a normal
> message *but* to do this, I had to change the normal-message payload
> length prefix to be 4 bytes. This only takes effect for version v1 and up
> though.
>
> I also added unit tests to ensure that the negotiations between different
> versions happens smoothly
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lightninglabs/lightning-node-connect/pull/26#issuecomment-1008889500>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHTWLWOGBORGDJDUVSMOJDUVLPP3ANCNFSM5JWJ3WKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ellemouton commented:
@Roasbeef & @guggero:
I have opened another PR using a different method so that it is easier to compare. 
- This PR has again been updated so that it sends the auth data in chunks in act 4
- The other PR #27 , sends the authdata  in a dynamic field in act 2
Please let me know which method I should go forward with.

lightninglabs-deploy commented:
@roasbeef: review reminder
@guggero: review reminder

lightninglabs-deploy commented:
@roasbeef: review reminder

ellemouton commented:
Closing this for now since it looks like we are going with the other approach
